File: src/mocks/handlers/tramites.ts
import { http, HttpResponse } from 'msw';
import { mockTramites } from '../data/tramites';
import { mockTiendas } from '../data/tiendas';
import { getUserFromRequest } from '../utils';

export const tramitesHandlers = [
  // GET /api/tramites — global list with filters
  http.get('*/api/tramites', async ({ request }) => {
    const url = new URL(request.url);
    const page = parseInt(url.searchParams.get('page') || '1');
    const pageSize = parseInt(url.searchParams.get('page_size') || '25');
    const search = url.searchParams.get('search')?.toLowerCase();
    const estado = url.searchParams.get('estado');
    const tipo = url.searchParams.get('tipo');
    const estadoGeo = url.searchParams.get('estado_geografico');
    const soloVencidos = url.searchParams.get('solo_vencidos') === 'true';
    const porVencerDias = url.searchParams.get('por_vencer_dias');

    let filtered = [...mockTramites];

    const user = getUserFromRequest(request);
    if (user?.rol === 'OPERATOR' && user.tiendas_asignadas) {
      filtered = filtered.filter((t) => user.tiendas_asignadas!.includes(t.tienda_id));
    }

    if (search) {
      filtered = filtered.filter(
        (t) =>
          t.nombre.toLowerCase().includes(search) ||
          (t.tienda_nombre?.toLowerCase().includes(search) ?? false),
      );
    }
    if (estado) filtered = filtered.filter((t) => t.estado === estado);
    if (tipo) filtered = filtered.filter((t) => t.tipo === tipo);
    if (soloVencidos) filtered = filtered.filter((t) => t.estado === 'vencido');
    if (porVencerDias) {
      const days = parseInt(porVencerDias);
      const limit = new Date();
      limit.setDate(limit.getDate() + days);
      filtered = filtered.filter((t) => {
        const venc = new Date(t.fecha_vencimiento);
        return venc <= limit && t.estado !== 'vencido';
      });
    }

    // Filter by geographic state (requires looking up the tienda)
    if (estadoGeo) {
      const tiendaIds = new Set(mockTiendas.filter((s) => s.estado === estadoGeo).map((s) => s.id));
      filtered = filtered.filter((t) => tiendaIds.has(t.tienda_id));
    }

    // Default sort: by expiration ascending
    filtered.sort(
      (a, b) => new Date(a.fecha_vencimiento).getTime() - new Date(b.fecha_vencimiento).getTime(),
    );

    const total = filtered.length;
    const totalPages = Math.ceil(total / pageSize);
    const start = (page - 1) * pageSize;
    let data = filtered.slice(start, start + pageSize);

    // Inject up-to-date documents to accurately reflect associations
    const { mockDocumentos } = await import('../data/documentos');
    data = data.map((t) => ({
      ...t,
      documentos: mockDocumentos.filter((d) => d.tramite_ids.includes(t.id)),
    }));

    return HttpResponse.json({ data, total, page, page_size: pageSize, total_pages: totalPages });
  }),

  // GET /api/tramites/:id
  http.get('*/api/tramites/:id', async ({ params }) => {
    const tramite = mockTramites.find((t) => t.id === params.id);
    if (!tramite) {
      return HttpResponse.json({ detail: 'Trámite no encontrado' }, { status: 404 });
    }

    const { mockDocumentos } = await import('../data/documentos');
    const upToDateDocs = mockDocumentos.filter((d) => d.tramite_ids.includes(tramite.id));

    // Enrich with mock observaciones and historial
    return HttpResponse.json({
      ...tramite,
      documentos: upToDateDocs,
      observaciones: [
        {
          id: `obs-${tramite.id}-1`,
          descripcion:
            'La fecha de vigencia del documento no coincide con la registrada en el sistema.',
          severidad: 'warning',
          fecha: new Date(Date.now() - 3 * 86400000).toISOString(),
        },
      ],
      historial: [
        {
          id: `hist-${tramite.id}-1`,
          entidad_tipo: 'tramite',
          entidad_id: tramite.id,
          accion: 'estado_cambiado',
          usuario_id: 'usr-001',
          usuario_nombre: 'Ana García López',
          fecha: new Date(Date.now() - 7 * 86400000).toISOString(),
          detalle: 'Estado cambiado a "En revisión"',
        },
        {
          id: `hist-${tramite.id}-2`,
          entidad_tipo: 'tramite',
          entidad_id: tramite.id,
          accion: 'documento_cargado',
          usuario_id: 'usr-002',
          usuario_nombre: 'Carlos Mendoza Ruiz',
          fecha: new Date(Date.now() - 14 * 86400000).toISOString(),
          detalle: 'Se cargó documento de respaldo',
        },
      ],
    });
  }),

  // PUT /api/tramites/:id
  http.put('*/api/tramites/:id', async ({ request, params }) => {
    const tramiteIndex = mockTramites.findIndex((t) => t.id === params.id);
    if (tramiteIndex === -1) {
      return HttpResponse.json({ detail: 'Trámite no encontrado' }, { status: 404 });
    }

    const updates = (await request.json()) as any;
    const tramite = mockTramites[tramiteIndex];

    // Update fields
    if (updates.nombre !== undefined) tramite.nombre = updates.nombre;
    if (updates.fecha_inicio !== undefined) tramite.fecha_inicio = updates.fecha_inicio;
    if (updates.fecha_vencimiento !== undefined)
      tramite.fecha_vencimiento = updates.fecha_vencimiento;
    if (updates.es_permanente !== undefined) {
      tramite.es_permanente = updates.es_permanente;
      if (tramite.es_permanente) {
        tramite.fecha_vencimiento = '';
        tramite.estado = 'vigente'; // Permanente doesn't expire
      }
    }

    // Calculate new status if not permanente
    if (!tramite.es_permanente && tramite.fecha_vencimiento) {
      const days = Math.ceil(
        (new Date(tramite.fecha_vencimiento).getTime() - new Date().getTime()) / (1000 * 3600 * 24),
      );
      if (days < 0) tramite.estado = 'vencido';
      else if (days <= 15) tramite.estado = 'por_vencer';
      else if (tramite.estado === 'vencido' || tramite.estado === 'por_vencer')
        tramite.estado = 'vigente';
    }

    // In a real app we'd add to historial here
    if (!tramite.historial) tramite.historial = [];
    tramite.historial.unshift({
      id: `hist-${tramite.id}-${Date.now()}`,
      entidad_tipo: 'tramite',
      entidad_id: tramite.id,
      accion: 'tramite_actualizado',
      usuario_id: 'usr-admin',
      usuario_nombre: 'Admin User',
      fecha: new Date().toISOString(),
      detalle: 'Detalles del trámite actualizados',
    });

    return HttpResponse.json(tramite);
  }),
];

File: backend/app/api/tramites.py
from fastapi import APIRouter

from app.api.deps import CurrentUser, DbSession
from app.core.pagination import PaginatedResponse, paginate
from app.schemas.tramite import Tramite, TramiteUpdate
from app.services import tramite_service

router = APIRouter()


def _serialize_tramite(t) -> dict:
    return {
        "id": t.id,
        "tienda_id": t.tienda_id,
        "tienda_nombre": getattr(t.tienda, "nombre", None)
        if hasattr(t, "tienda")
        else None,
        "nombre": t.nombre,
        "tipo": t.tipo,
        "estado": t.estado,
        "fecha_inicio": t.fecha_inicio.isoformat() if t.fecha_inicio else "",
        "fecha_vencimiento": t.fecha_vencimiento.isoformat()
        if t.fecha_vencimiento
        else "",
        "es_permanente": t.es_permanente,
        "es_recurrente": t.es_recurrente,
        "periodo_recurrencia": t.periodo_recurrencia,
        "asignado_a": t.asignado_a,
        "observaciones": [
            {
                "id": o.id,
                "descripcion": o.descripcion,
                "severidad": o.severidad,
                "fecha": o.fecha.isoformat(),
            }
            for o in getattr(t, "observaciones", [])
        ],
        "documentos": [
            {
                "id": d.id,
                "tramite_ids": [],
                "nombre_archivo": d.nombre_archivo,
                "url": d.url,
                "estado_ocr": d.estado_ocr,
                "datos_extraidos": d.datos_extraidos,
                "requiere_revision_manual": d.requiere_revision_manual,
                "cargado_por": d.cargado_por,
                "cargado_en": d.created_at.isoformat(),
                "tienda_id": d.tienda_id,
            }
            for d in getattr(t, "documentos", [])
        ],
        "historial": [],
    }


@router.get("", response_model=PaginatedResponse[Tramite])
async def list_tramites(
    db: DbSession,
    current_user: CurrentUser,
    page: int = 1,
    page_size: int = 25,
    search: str | None = None,
    estado: str | None = None,
    tipo: str | None = None,
    estado_geografico: str | None = None,
    solo_vencidos: bool = False,
    por_vencer_dias: int | None = None,
):
    items, total = await tramite_service.get_many(
        db,
        page=page,
        page_size=page_size,
        search=search,
        estado=estado,
        tipo=tipo,
        estado_geografico=estado_geografico,
        solo_vencidos=solo_vencidos,
        por_vencer_dias=por_vencer_dias,
        current_user=current_user,
    )
    return paginate([_serialize_tramite(t) for t in items], total, page, page_size)


@router.get("/{id}", response_model=Tramite)
async def get_tramite(db: DbSession, id: str, current_user: CurrentUser):
    t = await tramite_service.get_by_id(db, id, current_user=current_user)

    # We must fetch historial here
    from app.repositories import historial_repo

    hist = await historial_repo.get_by_entity(db, "tramite", t.id)
    serialized = _serialize_tramite(t)
    serialized["historial"] = [
        {
            "id": h.id,
            "entidad_tipo": h.entidad,
            "entidad_id": h.entidad_id,
            "accion": h.accion,
            "usuario_id": h.actor_id or "system",
            "usuario_nombre": "Sistema"
            if not h.actor_id
            else h.actor_id,  # Simplified for now
            "fecha": h.timestamp.isoformat(),
        }
        for h in hist
    ]
    return serialized


@router.patch("/{id}", response_model=Tramite)
async def update_tramite(
    db: DbSession, id: str, data: TramiteUpdate, current_user: CurrentUser
):
    t = await tramite_service.update(
        db, id, actor=current_user, **data.model_dump(exclude_unset=True)
    )
    return _serialize_tramite(t)
